File: tests/_deferred/java-gradle-junit5-selenium/src/test/java/pages/RegisterPage.java
package pages;

import config.TestConfig;
import io.qameta.allure.Step;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class RegisterPage extends BasePage {

    private static final By REGISTER_FORM = By.cssSelector("[data-testid='register-form']");
    private static final By LOGIN_INPUT = By.cssSelector("[data-testid='register-login-input']");
    private static final By PASSWORD_INPUT = By.cssSelector("[data-testid='register-password-input']");
    private static final By CONFIRM_PASSWORD = By.cssSelector("[data-testid='confirm-password-input']");
    private static final By SUBMIT = By.cssSelector("[data-testid='register-submit-button']");
    private static final By LOGIN_LINK = By.cssSelector("[data-testid='login-link']");

    public RegisterPage(WebDriver driver, TestConfig config) {
        super(driver, config);
    }

    @Step("Open register page")
    public RegisterPage openPage() {
        openPath("/register");
        return shouldBeOpen();
    }

    @Step("Click 'Login' link under the register form")
    public LoginPage clickLoginLink() {
        click(LOGIN_LINK);
        return new LoginPage(driver, config);
    }

    @Step("Fill and submit register form")
    public HomePage fillAndSubmitForm(String username, String password, String confirmPassword) {
        typeUsername(username);
        typePassword(password);
        typeConfirmPassword(confirmPassword);
        return submit();
    }

    @Step("Type username: {username}")
    public RegisterPage typeUsername(String username) {
        type(LOGIN_INPUT, username);
        return this;
    }

    @Step("Type password")
    public RegisterPage typePassword(String password) {
        type(PASSWORD_INPUT, password);
        return this;
    }

    @Step("Type confirm password")
    public RegisterPage typeConfirmPassword(String confirmPassword) {
        type(CONFIRM_PASSWORD, confirmPassword);
        return this;
    }

    @Step("Submit register form")
    public HomePage submit() {
        click(SUBMIT);
        return new HomePage(driver, config);
    }

    @Step("Verify register page is open")
    public RegisterPage shouldBeOpen() {
        waitVisible(REGISTER_FORM);
        return this;
    }
}

File: tests/python/tests-python-selene/pages/register_page.py
from __future__ import annotations

import allure
from selene import be, browser, have

from pages.base import BasePage


class RegisterPage(BasePage):
    @allure.step("Open register page")
    def open_page(self) -> RegisterPage:
        self.open_path("/register")
        return self.should_be_open()

    @allure.step("Click 'Login' link under the register form")
    def click_login_link(self):
        from pages.login_page import LoginPage

        browser.element("[data-testid='login-link']").should(be.visible).click()
        return LoginPage(self.config)

    @allure.step("Fill and submit register form")
    def fill_and_submit_form(self, username: str, password: str, confirm_password: str):
        self.type_username(username)
        self.type_password(password)
        self.type_confirm_password(confirm_password)
        return self.submit()

    @allure.step("Type username: {username}")
    def type_username(self, username: str) -> RegisterPage:
        browser.element("[data-testid='register-login-input']").set_value(username)
        return self

    @allure.step("Type password")
    def type_password(self, password: str) -> RegisterPage:
        browser.element("[data-testid='register-password-input']").set_value(password)
        return self

    @allure.step("Type confirm password")
    def type_confirm_password(self, confirm_password: str) -> RegisterPage:
        browser.element("[data-testid='confirm-password-input']").set_value(confirm_password)
        return self

    @allure.step("Submit register form")
    def submit(self):
        from pages.home_page import HomePage

        browser.element("[data-testid='register-submit-button']").click()
        return HomePage(self.config)

    @allure.step("Submit register form expecting validation error")
    def submit_expecting_error(self) -> RegisterPage:
        browser.element("[data-testid='register-submit-button']").click()
        browser.element("[data-testid='register-error-message']").should(be.visible)
        return self

    @allure.step("Verify error message: {message}")
    def should_have_error_message(self, message: str) -> RegisterPage:
        browser.element("[data-testid='register-error-message']").should(have.text(message))
        return self

    @allure.step("Verify register page is open")
    def should_be_open(self) -> RegisterPage:
        browser.element("[data-testid='register-form']").should(be.visible)
        return self

    @allure.step("Verify register form is mounted")
    def should_show_register_form(self) -> RegisterPage:
        browser.element("[data-testid='register-form-title']").should(be.visible)
        browser.element("[data-testid='register-login-input']").should(be.visible)
        browser.element("[data-testid='register-password-input']").should(be.visible)
        browser.element("[data-testid='confirm-password-input']").should(be.visible)
        browser.element("[data-testid='register-submit-button']").should(be.visible)
        return self

    @allure.step("Verify form title message: {message}")
    def should_have_form_title(self, message: str) -> RegisterPage:
        browser.element("[data-testid='register-form-title']").should(have.text(message))
        return self
